### DISCUSSION TITLE
Callable bounding behaviour

### DISCUSSION THREAD

**Discussion by tkukushkin** [microsoft/pyright#5607] [Q&A] [answered]
I found out a case, when pyright and mypy don't see any problems, but program fails at runtime:

```python
from typing import Callable

def dec(m: Callable[['Foo', int], int]) -> Callable[['Foo', str], int]:
    class Wrapper:
        def __call__(this, self: 'Foo', /, x: str) -> int:
            return m(self, int(x))
    return Wrapper()

class Foo:
    @dec
    def bar(self, x: int) -> int:
        return x + 1

print(Foo().bar('1'))
```

As far as I understand, python typing system doesn't have proper "Function" type and type checkers consider every `Callable` as function. Function is descriptor which makes it bounded method but `Callable` in general case may not be a descriptor.

In this case type checkers bound `Wrapper.__call__` and make method from it (remove `this` argument), but it is not how it works at runtime.

Do I understand it correctly? If yes, are there any discussions about it? It seems it is not big problem, but I wonder why this is so.

**Comment by erictraut** [comment, marked as answer]
Yes, this is a known limitation of the type system. There are some classes of runtime errors that will not be caught through static type checking.

Do you need a `Wrapper` class? It doesn't appear to be doing anything within your sample, but perhaps your actual use case is more complex. If you remove the class, it works fine:

```python
from typing import Callable

def dec(m: Callable[["Foo", int], int]) -> Callable[["Foo", str], int]:
    def wrapper(self: "Foo", /, x: str) -> int:
        return m(self, int(x))
    return wrapper

class Foo:
    @dec
    def bar(self, x: int) -> int:
        return x + 1

print(Foo().bar("1"))
```

**Reply by tkukushkin** [reply]
Yes, I don't need `Wrapper` as a class actually, I use it in this example only to show the problem.

So, do I understand correctly that this problem is not going to be resolved and there are no public discussions about it?

**Reply by erictraut** [reply]
You're welcome to search the [python/typing discussion forum](https://github.com/python/typing/discussions) and/or start a new discussion there.
